src/dir.rs

```diff
@@ -0,0 +1,272 @@
+    /// }
+    ///
+    /// // Temp directory should be deleted by now
+    /// assert_eq!(tmp_path.exists(), false);
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn path(&self) -> &path::Path {
+        self.path.as_ref().unwrap()
+    }
+
+    /// Unwraps the [`Path`] contained in the `TempDir` and
+    /// returns it. This destroys the `TempDir` without deleting the
+    /// directory represented by the returned `Path`.
+    ///
+    /// [`Path`]: http://doc.rust-lang.org/std/path/struct.Path.html
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::fs;
+    /// use tempfile::TempDir;
+    ///
+    /// # use std::io;
+    /// # fn run() -> Result<(), io::Error> {
+    /// let tmp_dir = TempDir::new()?;
+    ///
+    /// // Convert `tmp_dir` into a `Path`, destroying the `TempDir`
+    /// // without deleting the directory.
+    /// let tmp_path = tmp_dir.into_path();
+    ///
+    /// // Delete the temporary directory ourselves.
+    /// fs::remove_dir_all(tmp_path)?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn into_path(mut self) -> PathBuf {
+        self.path.take().unwrap()
+    }
+
+    /// Closes and removes the temporary directory, returing a `Result`.
+    ///
+    /// Although `TempDir` removes the directory on drop, in the destructor
+    /// any errors are ignored. To detect errors cleaning up the temporary
+    /// directory, call `close` instead.
+    ///
+    /// # Errors
+    ///
+    /// This function may return a variety of [`std::io::Error`]s that result from deleting
+    /// the files and directories contained with the temporary directory,
+    /// as well as from deleting the temporary directory itself. These errors
+    /// may be platform specific.
+    ///
+    /// [`std::io::Error`]: http://doc.rust-lang.org/std/io/struct.Error.html
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::fs::File;
+    /// use std::io::Write;
+    /// use tempfile::TempDir;
+    ///
+    /// # use std::io;
+    /// # fn run() -> Result<(), io::Error> {
+    /// // Create a directory inside of `std::env::temp_dir()`, named with
+    /// // the prefix, "example".
+    /// let tmp_dir = TempDir::new()?;
+    /// let file_path = tmp_dir.path().join("my-temporary-note.txt");
+    /// let mut tmp_file = File::create(file_path)?;
+    /// writeln!(tmp_file, "Brian was here. Briefly.")?;
+    ///
+    /// // By closing the `TempDir` explicitly we can check that it has
+    /// // been deleted successfully. If we don't close it explicitly,
+    /// // the directory will still be deleted when `tmp_dir` goes out
+    /// // of scope, but we won't know whether deleting the directory
+    /// // succeeded.
+    /// drop(tmp_file);
+    /// tmp_dir.close()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn close(mut self) -> io::Result<()> {
+        let result = remove_dir_all(self.path());
+
+        // Prevent the Drop impl from removing the dir a second time.
+        self.path = None;
```

Comment:
If the entire reason for wrapping the path in an `Option` is to prevent `Drop` from attempting to delete the directory a second time, then you should be able to eliminate the need for `Option` with this:

```rust
pub fn close(mut self) -> io::Result<()> {
    // Replace the path with an empty one to avoid a memory leak by mem::forget.
    let path = std::mem::replace(self.path, PathBuf::new());
    std::mem::forget(self); // Don't call drop
    remove_dir_all(&path)
}
```

The same would have to be done to `into_path()` of course.

This is cleaner than using an `Option` in my opinion.

Reply:
Another problem to consider is what to do if `close()` fails. The caller might wish to implement a retry if, for example, they've created a temporary directory on a network file system that might have intermittent failures. (I encounter this problem a lot with network shares under Windows.)

However, it looks like you want API compatibility with the `tempdir` crate, so you probably can't solve this problem.

Reply:
> The caller might wish to implement a retry if, for example, they've created a temporary directory on a network file system that might have intermittent failures. 

...

>  It looks like this problem has already been talked to death in

Also, in posix, one should generally *not* retry a close. The correct way to handle such errors is to catch them with an fsync first.

Reply:
@Stebalien Are you talking about `close(2)` or `remove_dir_all`? I agree about `close(2)`, but deleting files over a network can have intermittent failures where retrying is the right thing to do.

Reply:
Ah. I was thinking about close. You're right, this is a bit different. However, I kind of feel that this is, really, the best API. We could put the path into the error but that'll make this less ergonomic.